My Jetpack: put My Jetpack menu item at top of Jetpack menu section

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR puts My Jetpack menu item at top of Jetpack menu section. Also, when the site is disconnected, My Jetpack shows the connection screen allowing the user get reconnected to the site. No hard redirection.

Fixes #

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* My Jetpack: put My Jetpack menu item at top of Jetpack menu section

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Install and activate only one stand-alone plugin (e.g. Boost or Backup)
* Make sure the Jetpack menu is added and My Jetpack is the first item
* ~~Confirm that when you visit My Jetpack you are redirected to the interstitial page of the plugin you have active (e.g. `#/add-boost`)~~ Confirm that when you visit the My Jetpack page you see the regular dashboard with the product cards
* Activate a second stand-alone plugin
* Confirm that when you visit the My Jetpack page you see the regular dashboard with the product cards
* Confirm that adding products works from there
* Test differentt combinations and actions you think could be tricky

before | after
------|--------
<img width="333" alt="image" src="https://user-images.githubusercontent.com/77539/155235066-d68792ac-c4dd-47ec-bcff-d1812c839f4e.png"> | <img width="363" alt="image" src="https://user-images.githubusercontent.com/77539/155234981-f2de26f0-6962-498d-b04b-35dd10b39b46.png">

retrofox committed:
[not verified] my-jetpack: put My Jetpack menu item at top

retrofox committed:
[not verified] changelog

retrofox committed:
[not verified] my-jetpack: update version

oskosk commented:
We'll punt this for My jetpack 1.0. I just moved it to that milestone

kraftbj reviewed (dismissed):
Re-approved after bringing it up to date. The original PR had a version bump that was no longer needed.

leogermani commented:
Please let me have a look at this PR before merging. I want to test a few edge cases

leogermani reviewed (dismissed):
When we disconnect Jetpack from the My Jetpack page we have to redirect the user to the first menu item under the Jetpack top level menu.

Since Jetpack is disconnected, the first menu will no longer be My Jetpack. 

Now, we end up with an error. We have to fix this and get an updated information about where to redirect the user to.

Here's where this happens: https://github.com/Automattic/jetpack/blob/master/projects/packages/my-jetpack/_inc/hooks/use-my-jetpack-connection/index.js#L15

retrofox commented:
> When we disconnect Jetpack from the My Jetpack page we have to redirect the user to the first menu item under the Jetpack top level menu.

Because, by convention, the first menu item is the Jetpack plugin?

> Since Jetpack is disconnected, the first menu will no longer be My Jetpack.
> Now, we end up with an error. We have to fix this and get an updated information about where to redirect the user to.

What do you think about keeping the user in My Jetpack with a "User not connected" page?
Even more, we can use at least for now the `connection` page:

`https://testiing-site/wp-admin/admin.php?page=my-jetpack#/connection`

<img width="986" alt="image" src="https://user-images.githubusercontent.com/77539/156171637-e1506f8f-3165-417e-b57d-5328b4d37191.png">

leogermani commented:
> Because, by convention, the first menu item is the Jetpack plugin?

No, we redirect to whichever is the first menu item. If only Boost is active, will be Boost. If Boost and Backup are active, it will be one of the two...

> What do you think about keeping the user in My Jetpack with a "User not connected" page?

I think it makes sense, since we want to have the same experience for all plugins. This will need design, but we can start simpe and iterate

leogermani commented:
Let's make the My Jetpack page visible when there's no connection and then we can merge this

retrofox committed:
[not verified] my-jetpack: allow init even when no connected

retrofox committed:
[not verified] my-jetpack: redirect to #/connection when site not connected

retrofox committed:
[not verified] my-jetpack: propage registrationNonce to connection-screen

leogermani commented:
Ongoing discussion in p9dueE-4KC-p2

leogermani commented:
Considering the discussion we had in p9dueE-4KC-p2#comment-7644 , I think we can close this PR.

leogermani committed:
[not verified] handle stand-alone onboarding

On My Jetpack, if only one plugin is installed and site is disconnected
redirect the user to add this product.

leogermani committed:
[not verified] fix tests

leogermani committed:
[not verified] fix typo

retrofox reviewed (commented):
testing locally, I still see the interstitial page of backup even having backup and boost installed.

<img width="670" alt="image" src="https://user-images.githubusercontent.com/77539/161620986-902c5d98-cff4-43a4-9224-440b87e90a6a.png">

`productsWithActivePlugin` has only one item, which is the `backup`plugin.
Something related with the local dev-env?

leogermani commented:
I can't reproduce it... anything else different in your env?

![Uploading Captura de tela de 2022-04-04 17-07-24.png…]()

retrofox commented:
> [](https://github.com/Automattic/jetpack/pull/23052)I can't reproduce it... anything else different in your env?
> Uploading Captura de tela de 2022-04-04 17-07-24.png…

I can't see your image. Did you break githug? 
<img width="634" alt="image" src="https://user-images.githubusercontent.com/77539/161637486-8766eb26-9a6f-4945-b472-65a9501a8ae2.png">

retrofox commented:
> I can't reproduce it... anything else different in your env?

It doesn't work properly in my local dev-env. Probably because of how the dev-env handles the local plugins. It works was expected in a JN site

retrofox commented:
actually tested again, and it works as expected. there are two Boost plugins there
<img width="752" alt="image" src="https://user-images.githubusercontent.com/77539/161640730-d7b538e1-ccc7-43dd-be0d-a08d6d000b13.png">

retrofox commented on projects/packages/my-jetpack/_inc/hooks/use-products/index.js:
```diff
@@ -0,0 +1,54 @@
+	return {
+		productsThatRequiresUserConnection,
+		productsWithActivePlugin,
+		disconnectedAndOnePluginInstalled,
```
I wonder whether worths computing this `disconnectedAndOnePluginInstalled` here in the hook, because maybe it could be implemented straight to the component. Just thinking loud.

retrofox reviewed (commented):
It works as expected.
However, from my PoV, showing the interstitial page of the plugin when it's the only active one doesn't make sense. It's confusing. I'd expect to see the regular My Jetpack overview page.
For this reason, I delegate the approval to @oskosk and @keoshi .

Implementation wise, I left a simple suggestion but I think it's good enough to go. Other couple of eyes won't hurt, though.

keoshi commented:
First off, thanks for creating this PR! I have some thoughts and concerns about this approach, which I've openly shared before with all of your, but using this to document a few of the more tricky ones.

### Add product interstitial
When Boost (or Backup) is installed but not configured, I was not expecting to see the interstitial on the My Jetpack page. The primary action — “Add Jetpack Boost” — doesn't make sense because the plugin is already active, and it differs from the primary action on the Boost page: “Get started”.

We could do several things here, from changing the CTA, to align both views (interstitial/initial set up screen), from prioritizing the set up flow of the standalone plugin in lieu of showing My Jetpack, or even having an “empty view” for My Jetpack. My preference would be to rely on the plugin for the default flow, while showing as much of the My Jetpack page as we can.

<img width="1455" alt="image" src="https://user-images.githubusercontent.com/390760/161735158-f0c2d4eb-4f59-402f-81f1-2d6a5b4e3283.png">

<img width="1455" alt="image" src="https://user-images.githubusercontent.com/390760/161736212-5ebc8d8d-ae49-4970-bf5e-309cb1a5d985.png">

### Jetpack plugin

After installing and setting up Jetpack, we land on Recommendations, which is part of the “Dashboard” sidebar item (or used to be), but My Jetpack is highlighted instead. Detected several issues with this:

- A click on My Jetpack is required to actually see that page.
- Clicking Dashboard on the sidebar doesn't highlight it.
- Clicking Settings afterwards highlights the Dashboard item instead.
- My Jetpack link next to Recommendations is kind of obsolete with this change, and should probably be removed.

<img width="1455" alt="image" src="https://user-images.githubusercontent.com/390760/161733400-85bc9a65-373d-425d-b243-53b52d338b4c.png">

leogermani committed:
[not verified] Merge branch 'master' into update/my-jetpack-handle-menu-item

leogermani committed:
Do not redirect if not connected

leogermani commented:
I've refreshed the PR and changed it so it does not redirect the user anywhere. It simply displays My Jetpack as it is even if there is only one plugin active

jeherve commented:
Closing this for now because of the lack of activity on this. We can always reopen in the future if needed.
